libc/sem.cc

```diff
@@ -76,3 +105,72 @@ int sem_trywait(sem_t* s)
         return libc_error(EAGAIN);
     return 0;
 }
+
+static std::unordered_map<std::string, indirect_semaphore> named_semaphores;
+static mutex named_semaphores_mutex;
+
+OSV_LIBC_API
+sem_t *sem_open(const char *name, int oflag, mode_t mode, unsigned int value)
+{
+    SCOPE_LOCK(named_semaphores_mutex);
+    auto iter = named_semaphores.find(std::string(name));
+    
+    if (iter != named_semaphores.end()) {
+        //opening already named semaphore
+        if (oflag & (O_EXCL | O_CREAT)) {
+            errno = EEXIST;
+            return SEM_FAILED;
+        }
+
+        iter->second->add_reference();
+        return reinterpret_cast<sem_t*>(&(iter->second));
+    }
+    else if (oflag & O_CREAT) {
+        //creating new semaphore
+        if (value > SEM_VALUE_MAX) {
+            errno = EINVAL;
+            return SEM_FAILED;
+        }
+        /*
+         * Ignoring mode bits bits for now. Do we actually want to worry about 
+         * permissions in a unikernel? Would also need to check permissons when 
+         * using open and unlink if so.
+        */
+        named_semaphores.emplace(std::string(name), std::make_unique<posix_semaphore>(value, 1, true)); 
+        return reinterpret_cast<sem_t *>(&named_semaphores[std::string(name)]);
+    }
+    
+    errno = ENOENT;
+    return SEM_FAILED;
+}
+
+OSV_LIBC_API
+int sem_unlink(const char *name)
+{
+    SCOPE_LOCK(named_semaphores_mutex);
+    auto iter = named_semaphores.find(std::string(name));
+    if (iter != named_semaphores.end()) {
+        iter->second->unlink();
+        if (iter->second->not_referenced()) {
+            sem_destroy(reinterpret_cast<sem_t *>(&iter->second));
+        }
+        named_semaphores.erase(iter);
+        return 0;
+    }
+    
+    errno = ENOENT;
+    return -1;
+}
+
+OSV_LIBC_API
+int sem_close(sem_t *sem)
+{
+    SCOPE_LOCK(named_semaphores_mutex);
+    indirect_semaphore *named_sem = &from_libc(sem);                                  
+    
+    named_sem->get()->remove_reference();
```

Comment:
Nitpick: here and below, you don't need to use get() just to use a smart pointer. Unless I'm missing something, you can write `named_sem->remove_reference()` instead `named_sem->get()->remove_reference()`.

Reply:
The get was necessary the way I was using the `named_sem` object, which is a pointer to an `indirect_semaphore`. It isn't possible to use `indirect_semaphore named_sem = from_libc(sem);`, which I initially tried in order to avoid recasting the object, because the copy constructor for std::unique_ptr is deleted. An alternative option is to use `from_libc(sem)->` in place of all `named_sem->get()->`. Any preference?

Reply:
Oh, I missed that. Another thing that could work is, I think (didn't try) is to use `&` instead of `*`:
```c++
indirect_semaphore& named_sem = from_libc(sem);
named_sem->remove_reference();
```

Reply:
This does work and seems like the most concise option.